### PR TITLE
PPS: updates for re-miniAOD (backport of #29770)

### PR DESCRIPTION
**Pull request by jan-kaspar**
#### PR description:

~~This is currently a draft - tests are in progress on PPS side, expected to be concluded on Mon 11 May.~~ EDIT on 14 May: draft status removed, no more changes to this PR expected.

This PR contains **conditional** updates conceived for UL re-miniAOD campaign.

The updates are only activated when "run2_miniAOD_devel" flag is declared. The updates are intended to be used with updated conditions currently in tags CTPPSRPAlignment_real_offline_v8 + PPSOpticalFunctions_offline_v7.

When the "run2_miniAOD_devel" flag is not declared (and the current conditions are used), no change is produced wrt. the UL re-reco state.

#### PR validation:

The plot below shows a comparison of tests with old conditions and without the devel flag:
  * blue: before this PR
  * red dashed: with this PR

There's no difference.

make_cmp_10_6

The plot below shows a comparison of tests with updated conditions and with the devel flag:
  * blue: this PR
  * red: PR #29770

There is no difference.

make_cmp_interbranch

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

This is a backport of #29770. It is needed for the UL re-miniAOD campaign.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
backport of #29770

**Comment by slava77** [comment]
the addOnTest problem is from this PR
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-be920d/6182/addOnTests/fastsim2/cmsDriver.py_TTbar_13TeV_TuneCUETP8M1_cfi_--conditions_auto:run2_mc_--fast__-n_100_--eventcontent_AODSIM,DQM_--relval_100000,1000_-s_GEN,SIM,RECOBEFMI.log
```
   [4] Calling method for module TotemTimingRecHitProducer/'totemTimingRecHits'
```
this should not be happening in fastsim
IIUC this is a fastsim2 addOn
```
                  'fastsim2'  : ["cmsDriver.py TTbar_13TeV_TuneCUETP8M1_cfi --conditions auto:run2_mc --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot NominalCollision2015 --era Run2_2016"],
```

**Comment by jan-kaspar** [comment]
Thanks @slava77 . It looks like the PPS modules are not removed from fast sim (as they should be). I am investigating...

**Comment by jan-kaspar** [comment]
The problem was related to the transition Sequence -> Task. I've updated Configuration/StandardSequences/python/Reconstruction_cff.py as it is in master, see 8dd7982. I have rerun all the addOn tests and get
```
46 tests passed, 1 failed
```
The failing test is "pat1", log reading "Failed to open the file ... " Thus presumably unrelated to this PR.

Could someone restart the tests, please?

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by jfernan2** [comment]
+1

**Comment by jan-kaspar** [comment]
The PPS updates for re-miniAOD were discussed in Proton POG meeting:
https://indico.cern.ch/event/917867/
resulting in one more change of the near-far association cuts (da55959). At this moment, no more changes to this PR are expected. The draft/RFC status has been removed.

**Comment by jan-kaspar** [comment]
The plot below shows that with old conditions and without the devel flag there is no difference between:
  * 10_6 before this PR (blue)
  * the latest commit in this PR (red dashed)

**Comment by jan-kaspar** [comment]
The plot below shows that identical results are produced with updated conditions and the devel flag with:
  * this PR (blue)
  * #29770 (red dashed)

make_cmp_interbranch

**Comment by jan-kaspar** [comment]
Could someone trigger tests please?

**Comment by jfernan2** [comment]
please test

**Comment by jfernan2** [comment]
+1

File: RecoCTPPS/ProtonReconstruction/plugins/CTPPSProtonProducer.cc [hunk 1]
```diff
@@ -308,34 +341,37 @@ void CTPPSProtonProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSe
...
+              if (std::abs(z_i) >= std::abs(z_j))
+                continue;
+
               bool matching = true;
 
-              if (ac.x_cut_apply && std::abs(tr_i.getX() - tr_j.getX()) > ac.x_cut_value)
+              if (ac.x_cut_apply && std::abs(tr_i.getX() - tr_j.getX() - ac.x_cut_mean) > ac.x_cut_value)
                 matching = false;
-              if (ac.y_cut_apply && std::abs(tr_i.getY() - tr_j.getY()) > ac.y_cut_value)
+              if (ac.y_cut_apply && std::abs(tr_i.getY() - tr_j.getY() - ac.y_cut_mean) > ac.y_cut_value)
```
Comment (perrotta) [inline RecoCTPPS/ProtonReconstruction/plugins/CTPPSProtonProducer.cc:370]: ```else if```, as in the master

File: PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py [hunk 2]
```diff
@@ -548,13 +548,12 @@ def miniAOD_customizeOutput(out):
...
-    process.load("RecoCTPPS.ProtonReconstruction.ctppsProtons_cff")
-    process.load("Geometry.VeryForwardGeometry.geometryRPFromDB_cfi")
+    process.load("RecoCTPPS.Configuration.recoCTPPS_cff")
     task = getPatAlgosToolsTask(process)
     from Configuration.Eras.Modifier_ctpps_2016_cff import ctpps_2016
-    ctpps_2016.toModify(task, func=lambda t: t.add(process.ctppsLocalTrackLiteProducer))
-    ctpps_2016.toModify(task, func=lambda t: t.add(process.ctppsProtons))
+    from Configuration.Eras.Modifier_run2_miniAOD_devel_cff import run2_miniAOD_devel
+    (ctpps_2016 & ~run2_miniAOD_devel).toModify(task, func=lambda t: t.add(process.ctppsLocalTrackLiteProducer, process.ctppsProtons))
+    (ctpps_2016 & run2_miniAOD_devel).toModify(task, func=lambda t: t.add(process.recoCTPPSTask))
```
[the remark this answers is not in the document]
Comment (jan-kaspar) [reply PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py:556]: ctpps_2016 is actually applied also for 2017 and 2018, see e.g.
https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_X/Configuration/Eras/python/Era_Run2_2017_cff.py#L28
I know it is confusing, I don't like it either, but breaking the backward compatibility seemed to be even less good idea.

File: RecoCTPPS/ProtonReconstruction/plugins/CTPPSProtonProducer.cc (hunk 1 above)
Comment (jan-kaspar) [reply RecoCTPPS/ProtonReconstruction/plugins/CTPPSProtonProducer.cc:370]: Yep, will do it.

Comment (jan-kaspar) [reply RecoCTPPS/ProtonReconstruction/plugins/CTPPSProtonProducer.cc]: Done in 8b2d732.

**Comment by jan-kaspar** [comment]
The last commit, 8b2d732, is just a "cosmetic" change. The plot below shows that the results (with devel flag and updated conditions) are identical from this branch (blue) and master (red dashed).

make_cmp_interbranch

File: PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (perrotta) [reply PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py:556]: Correct, thank you Jan

**Comment by perrotta** [comment]
please test

**Comment by jfernan2** [comment]
+1

**Comment by perrotta** [comment]
+1
- PPS time and proton reconstrunction backported from the master, but those updates are only accessed with the run2_miniAOD_devel modifier.
- Validation plots verify that with that modifier the outputs are the same as in the master
- Jenkins tests show no differences for standard workflows, as it should

**Comment by santocch** [comment]
+1

**Comment by silviodonato** [comment]
@jan-kaspar : there are some changes that were not included in #29770. Could you make a PR to include these changes in master? I'm referring, for instance, to the conversion from `Sequence` of  `Task` of PPS sequence in `Configuration/StandardSequences/python/Reconstruction_cff.py`.

**Comment by jan-kaspar** [comment]
@silviodonato I am confused. The Sequence-to-Task changes in this PR were done following the code in master. Taking the example you point out, the PPS list in the copyAndExclude call should be identical in this PR:
  https://github.com/cms-sw/cmssw/pull/29771/files#diff-bc1d3eddbea8b75d590d62e72fee751eL93
and in master
  https://github.com/cms-sw/cmssw/blob/master/Configuration/StandardSequences/python/Reconstruction_cff.py#L96
Do I miss anything?

**Comment by silviodonato** [comment]
+operations
@jan-kaspar I was looking at the file changed by the PRs. Ok, so you are backporting also part of #28678

**Comment by silviodonato** [comment]
+1

**Pull request merged**